crates/fluvio/src/producer/mod.rs

```diff
@@ -33,12 +45,224 @@ use crate::producer::partitioning::{Partitioner, SiphashRoundRobinPartitioner, P
+                );
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Pool of producers for a given topic. There is a producer per partition
+struct ProducerPool {
+    pool: Vec<Arc<PartitionProducer>>,
+}
+
+impl ProducerPool {
+    fn new(
+        topic: String,
+        spu_pool: Arc<SpuPool>,
+        batches: Arc<HashMap<PartitionId, Arc<Mutex<VecDeque<ProducerBatch>>>>>,
+        linger: Duration,
+    ) -> Self {
+        let mut pool = vec![];
+        for (&partition_id, batch_list) in batches.iter() {
+            pool.push(PartitionProducer::shared(
+                topic.clone(),
+                partition_id,
+                spu_pool.clone(),
+                batch_list.clone(),
+                linger,
+            ));
+        }
+        Self { pool }
+    }
+
+    fn shared(
+        topic: String,
+        spu_pool: Arc<SpuPool>,
+        batches: Arc<HashMap<PartitionId, Arc<Mutex<VecDeque<ProducerBatch>>>>>,
+        linger: Duration,
+    ) -> Arc<Self> {
+        Arc::new(ProducerPool::new(topic, spu_pool, batches, linger))
+    }
+}
+struct InnerTopicProducer {
+    topic: String,
+    spu_pool: Arc<SpuPool>,
+    partitioner: Box<dyn Partitioner + Send + Sync>,
+    record_accumulator: RecordAccumulator,
+    producer_pool: Arc<ProducerPool>,
+    linger: Duration,
+}
+
+impl InnerTopicProducer {
+    /// Flush all the PartitionProducers and wait for them.
+    async fn flush(&self, force: bool) -> Result<()> {
+        let mut futures = vec![];
+
+        for producer in self.producer_pool.pool.iter() {
+            let future = producer.flush(force);
+            futures.push(future);
+        }
+
+        join_all(futures)
+            .await
+            .into_iter()
+            .collect::<Result<Vec<_>>>()?;
+
+        Ok(())
+    }
+
+    async fn push_record(self: Arc<Self>, record: Record) -> Result<PushRecord> {
+        let topics = self.spu_pool.metadata.topics();
+        let topic_spec = topics
+            .lookup_by_key(&self.topic)
+            .await?
+            .ok_or_else(|| FluvioError::TopicNotFound(self.topic.to_string()))?
+            .spec;
+        let partition_count = topic_spec.partitions();
+        let partition_config = PartitionerConfig { partition_count };
+
+        let key = record.key.as_ref().map(|k| k.as_ref());
+        let value = record.value.as_ref();
+        let partition = self.partitioner.partition(&partition_config, key, value);
+
+        let push_record = self
+            .record_accumulator
+            .push_record(record, partition)
+            .await?;
+        let self_clone = self.clone();
+        if push_record.is_full {
+            fluvio_future::task::spawn(async move {
+                if let Err(e) = self_clone.flush(false).await {
+                    error!("Failed to flush producer: {:?}", e);
+                }
+            });
+        }
+
+        if push_record.new_batch {
```

Comment:
what does mean new batch?

Reply:
So, in the RecordAccumulator, we have a VecDeq of batches per partition, each time a batch is full (due to max size constraints, we create a new batch in the VecDeq and add the record there in the new batch. Every time it is created a new batch we need to spawn a flush once the linger time is reached.

This implementation is pretty much based on this 

https://github.com/apache/kafka/blob/394175688a76d52d1895c4419711edb4b47bfd47/clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java#L180

Reply:
see other comments about lingering.  Kafka uses traditional thread-based approach while we use async approach which requires different implementation strategy